src/devices/cpu/i386/i386segs.hxx

```diff
@@ -2490,48 +2494,51 @@ inline void i386_device::dri_changed()
 			int breakpoint_length = (m_dr[7] >> ((dr << 2) + 16 + 2)) & 3;
 			uint32_t phys_addr = m_dr[dr];
 			uint32_t error;
-			phys_addr = translate_address(m_CPL, TRANSLATE_READ, &phys_addr, &error);
-			phys_addr &= ~3; // According to CUP386, data breakpoints are only reliable on dword-aligned addresses, so align this to a dword.
-			uint32_t true_mask = 0;
-			switch(breakpoint_length)
+			if(translate_address(m_CPL, TRANSLATE_READ, &phys_addr, &error))
 			{
-				case 0: true_mask = 0xff; break;
-				case 1: true_mask = 0xffff; break;
-				// Case 2 is invalid on a real 386.
-				case 3: true_mask = 0xffffffff; break;
-			}
-			if(true_mask == 0)
-			{
-				logerror("i386: Unknown breakpoint length value\n");
-			}
-			else if(breakpoint_type == 1) m_dr_breakpoints[dr] = m_program->install_write_tap(phys_addr, phys_addr + 3, "i386_debug_write_breakpoint",
-			[&, dr, true_mask](offs_t offset, u32& data, u32 mem_mask)
-			{
-				if(true_mask & mem_mask)
+				phys_addr &= ~3; // According to CUP386, data breakpoints are only reliable on dword-aligned addresses, so align this to a dword.
+				uint32_t true_mask = 0;
+				switch(breakpoint_length)
 				{
-					m_dr[6] |= 1 << dr;
-					i386_trap(1,0,0);
+					case 0: true_mask = 0xff; break;
+					case 1: true_mask = 0xffff; break;
+					// Case 2 is invalid on a real 386.
+					case 3: true_mask = 0xffffffff; break;
 				}
-			}, &m_dr_breakpoints[dr]);
-			else if(breakpoint_type == 3) m_dr_breakpoints[dr] = m_program->install_readwrite_tap(phys_addr, phys_addr + 3, "i386_debug_readwrite_breakpoint",
-			[this, dr, true_mask](offs_t offset, u32& data, u32 mem_mask)
-			{
-				if(true_mask & mem_mask)
+				if(true_mask == 0)
 				{
-					m_dr[6] |= 1 << dr;
-					i386_trap(1,0,0);
+					logerror("i386: Unknown breakpoint length value\n");
 				}
-			},
-			[this, dr, true_mask](offs_t offset, u32& data, u32 mem_mask)
-			{
-				if(true_mask & mem_mask)
+				else if(breakpoint_type == 1) m_dr_breakpoints[dr] = m_program->install_write_tap(phys_addr, phys_addr + 3, "i386_debug_write_breakpoint",
+				[&, dr, true_mask](offs_t offset, u32& data, u32 mem_mask)
```

Comment:
It’s not safe to capture *anything* by reference when a lambda will outlive its containing scope.  Also, auto capture is a bug waiting to happen.  Please capture what you need explicitly.

Also, please clean up the formatting here – if the controlled statement spans multiple lines, it’s cleared if you use braces, and statement continuations should be intended (I realise it was already a mess.  Make is something like:
```
            else if(breakpoint_type == 1)
            {
                m_dr_breakpoints[dr] = m_program->install_write_tap(
                        phys_addr,
                        phys_addr + 3,
                        "i386_debug_write_breakpoint",
                        [...] (offs_t offset, u32 &data, u32 mem_mask)
                        {
                            ...
                        },
                        &m_dr_breakpoints[dr]);
            }
```